docs: Docs nav structure reorg follow-up

aaronsteers opened the pull request:
I did a pass to clean up titles and organization a little bit, following from https://github.com/meltano/sdk/pull/1353#issuecomment-1414272456.

Preview:

- https://meltano-sdk--1407.org.readthedocs.build/en/1407/

Notable changes:

Removed "SDK Implementation Details - " as a prefix from pages in this section:

<img width="242" alt="image" src="https://user-images.githubusercontent.com/18150651/217685023-f675e77f-88e8-488c-92fe-de1e6c8664f5.png">

Moved "Plugin Capabilities" to Reference:

<img width="243" alt="image" src="https://user-images.githubusercontent.com/18150651/217685079-88e96585-9c05-40d2-af31-d71d2c798efe.png">

Combined "Contributing" page and "Release Process" under common parent. Raised "Deprecation Timeline" up one level to be visible to non-contributors/non-maintainers.

<img width="238" alt="image" src="https://user-images.githubusercontent.com/18150651/217685154-2ece9db6-6c51-4344-a1c9-30c17dbe893e.png">

WillDaSilva commented on docs/CONTRIBUTING.md:
```diff
@@ -1,4 +1,4 @@
-# Contributing to the SDK
+# Contributor's Guide
```
```suggestion
# Contributing Guide
```

WillDaSilva commented on docs/development/release_process.md:
```diff
@@ -2,7 +2,7 @@
 ## PyPI releases
 
-TODO: This section needs to be updated to reflect the release process.
+Releases are published to PyPi by a GitHub Actions workflow, which in turn is triggered by a GitHub [Release](https://github.com/meltano/sdk/releases) being published.
```
```suggestion
Releases are published to PyPI by a GitHub Actions workflow, triggered when a GitHub [Release](https://github.com/meltano/sdk/releases) is published.
```

WillDaSilva commented on docs/index.rst:
```diff
@@ -68,16 +69,15 @@ within the `#singer-tap-development`_ and `#singer-target-development`_ Slack ch
    :maxdepth: 1
    :hidden:
 
-   CONTRIBUTING
```
Why remove this?

aaronsteers replied:
Moved under "development" parent.

<img width="240" alt="image" src="https://user-images.githubusercontent.com/18150651/217685495-59a04d29-1965-477e-a7fd-426bd5aed76a.png">

If we raise up, we should raise up "Release Process" also; otherwise the "Development" parent would only have a single parent.

aaronsteers replied:
Optionally, we could just have "Contributor's Guide", "Release Process", and "Deprecation Timeline" as top-level items under the "SDK Development" section.

@edgarrmondragon - any objections to us moving them up to the top level?

aaronsteers replied:
As noted in the description, I do think "Deprecation Timeline" should be visible to non-contributors, since that effects tap/target maintainers even if they are not themselves contributors.

WillDaSilva replied:
Oh I see where it is now. I was just failing to see the new location.

I have a slight preference for moving it to the top-level under "SDK Development". There's only 2 items nested there after all, and the name they're nested under is identical to the category name above.

edgarrmondragon replied:
> any objections to us moving these up to the top level?

@aaronsteers no objections, it makes sense :+1:

WillDaSilva reviewed (approved):
Looks good to me. Moving those 2 items out from "SDK Development" would be nice, but this PR is still good with or without that.

edgarrmondragon reviewed (approved):
This LGTM, regardless if those two pages are moved up.

aaronsteers commented:
I noticed we removed the dedicated `typing` page when consolidating into `classes`. Because there was helpful docstring at the top of that module, and because I like having a clean discoverable top-level nav for this, I've reverted the removal of the other page. I added `:noindex:` to the restored page so that the new pages are still the authority for class lookups.
